app/controllers/market_controller.rb

```diff
@@ -171,45 +173,43 @@ def create_buy
         type = "item"
         name = name[/\(.*?\)/].gsub("(", "").gsub(")", "")
       else
-        render(json: { 'error_message': I18n.t('errors.name_not_found') }, status: 400) && (return)
+        render(json: { 'error_message': I18n.t('errors.name_not_found') }, status: :bad_request) && (return)
       end
 
       MarketListing.create(loader: name, listing_type: type, location: current_user.location, price: price, amount: amount, user: current_user, order_type: :buy)
 
-      render(json: {}, status: 200) && (return)
+      render(json: {}, status: :ok) && (return)
     end
-    render json: {}, status: 400
+    render json: {}, status: :bad_request
   end
 
   def fulfill_buy
     if params[:id] && params[:amount]
       amount = params[:amount].to_i rescue nil
-      listing = MarketListing.find(params[:id]) rescue nil
+      listing = MarketListing.ensure(params[:id])
       if amount && listing && (listing.location == current_user.location) && (amount >= 1)
         # Check Amount
-        render(json: { 'error_message': I18n.t('errors.you_dont_have_enough_of_this') }, status: 400) && (return) if (Item.find_by(loader: listing.loader, location: current_user.location, user: current_user).count rescue 0) < amount
-        render(json: { 'error_message': I18n.t('errors.buyer_doesnt_want_that_much') }, status: 400) && (return) if amount > listing.amount
+        render(json: { 'error_message': I18n.t('errors.you_dont_have_enough_of_this') }, status: :bad_request) && (return) if (Item.find_by(loader: listing.loader, location: current_user.location, user: current_user).count rescue 0) < amount
+        render(json: { 'error_message': I18n.t('errors.buyer_doesnt_want_that_much') }, status: :bad_request) && (return) if amount > listing.amount
         # Remove Items and give credits
-        Item.remove_from_user(loader: listing.loader, amount: amount, location: current_user.location, user: current_user)
+        Item::RemoveFromUser.(loader: listing.loader, amount: amount, location: current_user.location, user: current_user)
         current_user.give_units(listing.price * amount * 0.95)
         # Give Items to Buyer and reduce listing
-        Item.give_to_user(loader: listing.loader, amount: amount, location: current_user.location, user: listing.user)
+        Item::GiveToUser.(loader: listing.loader, amount: amount, location: current_user.location, user: listing.user)
         new_amount = listing.amount - amount
-        listing.update_columns(amount: new_amount)
+        listing.update(amount: new_amount)
```

Comment:
General Question about changing `update_columns` to `update` : Wouldn't update_columns be a lot faster because we skip callbacks and validations here? I think its okay for everything we can control and will keep the game fast. Your opinion?

Reply:
in my experience, all the safeguards that `update_columns` bypasses are rarely the speed problem.   you also end up `reload`-ing models a lot because the db has changed underneath, and run the risk of bad data because validations have been skipped.